parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/HadoopStreams.java

```diff
@@ -0,0 +1,116 @@
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.apache.parquet.hadoop.util;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.parquet.Log;
+import org.apache.parquet.io.ParquetDecodingException;
+import org.apache.parquet.io.SeekableInputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.nio.ByteBuffer;
+
+public class HadoopStreams {
+
+  private static final Log LOG = Log.getLog(SeekableInputStream.class);
+
+  private static final Class<?> byteBufferReadableClass;
+
+  private static final Constructor<SeekableInputStream> h2SeekableConstructor;
+
+  static {
+    byteBufferReadableClass = getReadableClass();
+    h2SeekableConstructor = getH2SeekableConstructor();
+  }
+
+  public static SeekableInputStream wrap(FSDataInputStream stream) {
+    if (byteBufferReadableClass != null && h2SeekableConstructor != null &&
+        byteBufferReadableClass.isInstance(stream) &&
+        supportsByteBufferReads(stream)) {
+      try {
+        return h2SeekableConstructor.newInstance(stream);
+      } catch (InstantiationException e) {
+        LOG.warn("Could not instantiate H2SeekableInputStream, falling back to byte array reads", e);
+        return new H1SeekableInputStream(stream);
+      } catch (IllegalAccessException e) {
+        LOG.warn("Could not instantiate H2SeekableInputStream, falling back to byte array reads", e);
+        return new H1SeekableInputStream(stream);
+      } catch (InvocationTargetException e) {
+        throw new ParquetDecodingException(
+            "Could not instantiate H2SeekableInputStream", e.getTargetException());
+      }
+    } else {
+      return new H1SeekableInputStream(stream);
+    }
+  }
+
+  private static final ByteBuffer ZERO_LEN_BYTE_BUFFER = ByteBuffer.wrap(new byte[0]);
+
+  private static boolean supportsByteBufferReads(FSDataInputStream stream) {
+    // FSDataInputStream implements ByteBufferReadable, but may throw
+    // UnsupportedOperationException if it isn't actually supported. This tests
+    // whether the method throws by trying to read a 0-length buffer, which has
+    // no effect on the stream when it is supported.
+    try {
+      stream.read(ZERO_LEN_BYTE_BUFFER);
+      return true;
+    } catch (UnsupportedOperationException e) {
+      return false;
```

Comment:
to help debugging should we log the false / null scenarios in this class? we basically end up falling back in them to the HadoopV1 stream. We could log as debug as it might be spammy